src/lib/FreeC/Pass/SharingAnalysisPass.hs

```diff
@@ -0,0 +1,144 @@
+-- | This module contains a compiler pass that analyses the right-hand sides of
+--   function declaration and introduces @let@ expression with new variables
+--   for each variable that occurs more than once on the right-hand sides.
+--
+--   = Examples
+--
+--   == Example 1
+--
+--   > twice :: Integer -> Integer
+--   > twice (x :: Integer) = x + x
+--
+--   Should be transformed into
+--
+--   > twice :: Integer -> Maybe Integer
+--   > twice (x :: Integer) = let (y :: Integer) = x in y + y
+--
+--   == Example 2
+--
+--   > twiceMaybe :: Maybe Integer -> Maybe Integer
+--   > twiceMaybe (mx :: Maybe Integer) = case (mx :: Maybe Integer) of {
+--   >     Nothing -> Nothing;
+--   >     Just x  -> Just (x + x)
+--   >   }
+--
+--   Should be transformed into
+--
+--   > twiceMaybe :: Maybe Integer -> Maybe Integer
+--   > twiceMaybe (mx :: Maybe Integer) = case (mx :: Maybe Integer) of {
+--   >     Nothing -> Nothing;
+--   >     Just x  -> let y = x in Just (y + y)
+--   >   }
```

Comment:
If I understand the code correctly, this is not implemented yet.
However, I'm also not convinced that this is semantically correct. Consider the following program.

```haskell
let m42 = Just (trace "Test" 42) in (twiceMaybe m42, twiceMaybe m42)
```

This program should print "Test" only once but I think that it will be printed twice in the example because `y` is shared only locally.

Reply:
You are right. I think misunderstood the task here.

Reply:
You are right. I overlooked to implement introduction of `let`-expressions on right hand sides of `case`-alternatives.

Reply:
@just95 If I understand your example correctly, the `let m42` binding introduces sharing, which is not "overwritten" by the binding in `twiceMaybe`. So it does not make a difference whether there is a `let` clause in the right-hand side of the `Just` alternative.

Reply:
@nbun Ah, okay, I somehow thought `share` would have no effect on the computation inside `Just`. I've now played around with the Coq implementation and can confirm that "Test" is printed only once in my example.